Add instructions against common AI poor code quality habits (#718)
vllm-project/guidellm: AGENTS.md

- Use appropriate markers (`smoke`, `sanity`, `regression`)
- Tests should be placed in files matching the name and path of the file under tests. E.g. `src/guidellm/benchmark/schemas/generative/entrypoints.py` -> `tests/unit/benchmark/schemas/generative/test_entrypoints.py`.

### Quality Requirements

- All Python code must pass linting and formatting
- All Python code must pass type checking
- All tests must pass before committing
- Markdown files must be properly formatted

### Style Requirements

- Public functions in `src/` code must use the reStructuredText docstring format
- All imports **SHALL** be done at the top of the file
- **DO NOT** use `getattr` or `setattr` as it hides incorrect usage of types

### Design Requirements

- Only touch sections of code that need to be changed for the given task

> - Prefer solutions that keep non-generic code out of generic pathways

Reviewer: What does this mean? I am not entirely sure what this entails so idk if an AI would figure it out.

Author: How about this?

> Do not add if/else branches for variant-specific logic in generic code paths. Variant-specific logic belongs in methods on registry class implementations

Or
> When handling variant-specific logic, encapsulate it in methods on registry class implementations rather than adding if/else branches to generic code paths

Reviewer 2: Or -- class implementations must fully encapsulate their unique logic and that logic must not leak into caller code paths.

Reviewer: I like @jaredoconnell's option 2 and @dbutenhof's. Maybe add both?

- All tests must pass before committing
- Markdown files must be properly formatted
- Public functions in `src/` code must use the reStructuredText docstring format

> - All imports must be done at the top of the file unless necessary for functionality

Reviewer 2: Can we be more specific? I think the only time we legitimately do that is to narrow the scope of optional extra packages; so maybe even go very specific like "unless necessary to minimize the scope of optional imports listed under the `pyproject.toml` "project.optional-dependencies" section."

Reviewer: I would make this more broad, not more specific. If you give AI an excuse to ignore a rule it will twist its logic to match.

```suggestion
- All imports **SHALL** be done at the top of the file
```

> - Use of `getattr` should be avoided if possible as it hides incorrect usage of types

Reviewer 2: We use `getattr` quite a bit -- again, I wonder if we can be more specific here. I kinda worry about how an over-eager-to-please LLM might choose to interpret "if possible". I'm not really coming up with a better condition right now, but this feels like a plot hole ... 😆

Reviewer: This is general guidance for AIs, not humans. If a human wants to override this guidance for a good reason they are free to do so.

```suggestion
- **DO NOT** use `getattr` or `setattr` as it hides incorrect usage of types
```

Reviewer 2: My concern is that an over-eager AI assistant will notice & decide to "fix" existing instances. Ah well -- we can burn that bridge when we reach it.

Author: I added a design requirements section that should address the over-eager changes.
